brave/src/main/java/brave/propagation/B3SingleFormat.java

```diff
@@ -158,79 +171,98 @@ public static TraceContextOrSamplingFlags parseB3SingleFormat(CharSequence value
+          }
+        }
 
-    long spanId = tryParseHex("span ID", value, pos, endIndex);
-    if (spanId == 0) return null;
-    pos += 16; // spanid
+        if (!validateFieldLength(currentField, currentFieldLength)) {
+          return null;
+        }
 
-    int flags = 0;
-    long parentId = 0L;
-    if (hyphenCount > 1) { // traceid-spanid-
-      pos++; // consume the hyphen
-
-      if (hyphenCount == 3) { // we should parse sampled AND parent ID
-        flags = tryParseSampledFlags(value, endIndex, pos);
-        if (flags == 0) return null;
-        pos += 2; // consume the sampled flag and hyphen
-        parentId = tryParseParentId(value, endIndex, pos);
-        if (parentId == 0L) return null;
-      } else { // we should parse sampled OR parent ID
-        if (endIndex - pos <= 1) {
-          flags = tryParseSampledFlags(value, endIndex, pos);
-          if (flags == 0) return null;
+        if (currentField.required && buffer == 0L) {
+          Platform.get().log("Invalid input: read all zeroes {0}", currentField.name, null);
+          return null;
+        }
+
+        if (currentField.equals(Field.TRACE_ID)) {
+          traceId = buffer;
+
+          currentField = Field.SPAN_ID;
+        } else if (currentField.equals(Field.SPAN_ID)) {
+          spanId = buffer;
+
+          // The handle malformed cases like below, it is easier to assume the next field is sampled
+          // and revert if definitely not vs try to determine if it is definitely sampled.
+          // 'traceId-spanId--parentSpanId'
+          // 'traceId-spanId-'
+          currentField = Field.SAMPLED;
+        } else if (currentField.equals(Field.SAMPLED)) {
+          SamplingFlags samplingFlags = tryParseSamplingFlags(value.charAt(pos - 1));
+          if (samplingFlags == null) return null;
+          flags = samplingFlags.flags;
+
+          currentField = Field.PARENT_SPAN_ID;
+          currentFieldLength = 0;
+          continue;
         } else {
-          parentId = tryParseParentId(value, endIndex, pos);
-          if (parentId == 0L) return null;
+          parentId = buffer;
+
+          if (!isEof) {
+            Platform.get().log("Invalid input: more than 4 fields exist", null);
+            return null;
+          }
         }
+
+        buffer = 0L;
+        currentFieldLength = 0;
+        continue;
+      } else if (currentField == Field.TRACE_ID && beginIndex + 16 == pos) {
+        // special casing the only valid non-hyphen at position 16
+        // we don't guard against all zeros as traceIdHigh can be all zeros for 128-bit
+        traceIdHigh = buffer;
+
+        // This character is the next hex. If it isn't, the next iteration will throw. Either way,
+        // reset so that we can capture the next 16 characters of the trace ID.
+        buffer = 0;
+        currentFieldLength = 0;
       }
+
+      // The rest of this is normal lower-hex decoding
+      buffer <<= 4;
+      if (c >= '0' && c <= '9') {
```

Comment:
Realize this is the same as current, but if pursuing more benchmarking, maybe worth looking into either having a switch with sixteen cases or a lookup table.

Reply:
ooo I like

Reply:
running benchmarks again now. I did all the other tricks I could think of along the way